images/daemon/pkg/daemon/kata_openshift.go

```diff
@@ -0,0 +1,459 @@
+
+	nodeName, err := getNodeName()
+	if err != nil {
+		return err
+	}
+
+	if !isKataUnInstalled {
+		// Kata binaries need to be uninstalled
+		err = updateKataConfigStatus(k.KataClient, kataConfigResourceName, func(ks *kataTypes.KataConfigStatus) {
+			ks.UnInstallationStatus.InProgress.InProgressNodesCount++
+		})
+
+		if err != nil {
+			return fmt.Errorf("kata is not installed on the node, error updating kataconfig status %+v", err)
+		}
+
+		if k.KataBinaryUnInstaller == nil {
+			k.KataBinaryUnInstaller = uninstallRPMs
+		}
+
+		err = k.KataBinaryUnInstaller(k)
+
+		if err != nil {
+			// kata uninstallation failed. report it.
+			err = updateKataConfigStatus(k.KataClient, kataConfigResourceName, func(ks *kataTypes.KataConfigStatus) {
+				ks.UnInstallationStatus.InProgress.InProgressNodesCount--
+
+				fn, err := getFailedNode(err)
+				if err != nil {
+					return
+				}
+
+				ks.UnInstallationStatus.Failed.FailedNodesList = append(ks.UnInstallationStatus.Failed.FailedNodesList, fn)
+				ks.UnInstallationStatus.Failed.FailedNodesCount = len(ks.UnInstallationStatus.Failed.FailedNodesList)
+			})
+
+			if err != nil {
+				return fmt.Errorf("kata installation failed, error updating kataconfig status %+v", err)
+			}
+
+		}
+		// mark binaries uninstalled
+		err = updateKataConfigStatus(k.KataClient, kataConfigResourceName, func(ks *kataTypes.KataConfigStatus) {
+			ks.UnInstallationStatus.InProgress.BinariesUnInstalledNodesList = append(ks.UnInstallationStatus.InProgress.BinariesUnInstalledNodesList, nodeName)
+		})
+
+		if err != nil {
+			return fmt.Errorf("kata uninstallation succeeded, but error updating kataconfig status %+v", err)
+		}
+	}
+
+	return nil
+}
+
+func doCmd(cmd *exec.Cmd) error {
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err := cmd.Run()
+	fmt.Println(cmd.String())
+	if err != nil {
+		log.Println(err)
+		return err
+	}
+
+	return nil
+}
+
+func rpmostreeOverrideReplace(rpms string) error {
+	cmd := exec.Command("/bin/bash", "-c", "/usr/bin/rpm-ostree override replace /opt/kata-install/packages/"+rpms)
+	err := doCmd(cmd)
+	if err != nil {
+		fmt.Println("override replace of " + rpms + " failed")
+		return err
+	}
+	return nil
+}
+
+func cleanupHost() error {
+	cmd := exec.Command("/usr/bin/rm", "-rf", "/opt/kata-install")
+	err := doCmd(cmd)
+	if err != nil {
+		return err
+	}
+
+	cmd = exec.Command("/usr/bin/rm", "-rf", "/usr/local/kata")
+	err = doCmd(cmd)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func uninstallRPMs(k *KataOpenShift) error {
+	log.SetOutput(os.Stdout)
+
+	if err := syscall.Chroot("/host"); err != nil {
+		log.Fatalf("Unable to chroot to %s: %s", "/host", err)
+	}
+
+	if err := syscall.Chdir("/"); err != nil {
+		log.Fatalf("Unable to chdir to %s: %s", "/", err)
+	}
+
+	err := cleanupHost()
+	if err != nil {
+		log.Println("cleanupHost failed")
+	}
+
+	cmd := exec.Command("rpm-ostree", "uninstall", "--idempotent", "--all") //FIXME not -a but kata-runtime, kata-osbuilder,...
```

Comment:
What prevents this comment `//FIXME not -a but kata-runtime, kata-osbuilder,...` from being applied in this PR?

Reply:
Yes, this should be fixed. It hasn't been a problem so far because no one else has installed additional RPMs on the nodes, but to make sure we don't uninstall anything that wasn't intended we should explicitly uninstall kata-runtime, kata-osbuilder RPMs. I'll open an issue for it and fix it in a follow-up PR once this is merged

Reply:
opened issue #29